DRODLib/Character.cpp

```diff
@@ -5324,17 +5324,113 @@ void CCharacter::Disappear()
 }
 
 //*****************************************************************************
-int CCharacter::GetIndexOfCommandWithLabel(const UINT label) const
+int CCharacter::GetIndexOfCommandWithLabel(const int label) const
 //Returns: index of command with specified label, or NO_LABEL if none
 {
-	if (label)
-		for (UINT wIndex=this->commands.size(); wIndex--; )
+	if (label > 0) {
+		for (UINT wIndex = this->commands.size(); wIndex--; )
 		{
 			const CCharacterCommand& command = this->commands[wIndex];
 			if (command.command == CCharacterCommand::CC_Label &&
-					label == command.x)
+				label == command.x)
 				return wIndex;
 		}
+	} else if (label < 0) {
+		const ScriptFlag::GotoSmartType eGotoType = (ScriptFlag::GotoSmartType) label;
+		switch (eGotoType) {
+		case ScriptFlag::GotoSmartType::PreviousIf:
+			return GetIndexOfPreviousIf(true);
+		break;
+		case ScriptFlag::GotoSmartType::NextElseOrElseIfSkipCondition:
+		{
+			int wIndex = GetIndexOfNextElse(false);
+			if (wIndex != NO_LABEL) {
+				return wIndex + 2;
+			}
+		}
+		break;
+		default:
+			return NO_LABEL;
+		break;
+		}
+	}
+	return NO_LABEL;
+}
+//*****************************************************************************
+//Return: Index of the if or else if command at the beginning of the block or
+// NO_LABEL if not in a block
+int CCharacter::GetIndexOfPreviousIf(const bool bIgnoreElseIf) const
+{
+	UINT wCommandIndex = this->wCurrentCommandIndex;
+	UINT wNestingDepth = 0;
+	bool bScanning = true;
+
+	do {
+		--wCommandIndex;
+		if (wCommandIndex == -1) //No matching if found
+			break;
+
+		CCharacterCommand command = this->commands[wCommandIndex];
+		switch (command.command) {
+		case CCharacterCommand::CC_If:
```

Comment:
Ditto with indent.
Sidenote: why not just return `wCommandIndex`, remove `bScanning` completely and move the above `if (...) break` to the while's condition?